11-designing-for-mental-health/projects/final/app/src/main/java/com/raywenderlich/android/tacotuesday/onboarding/OnboardingActivity.kt

```diff
@@ -0,0 +1,145 @@
+import com.raywenderlich.android.tacotuesday.databinding.ItemOnboardingBinding
+
+class OnboardingActivity : FragmentActivity() {
+
+  private lateinit var binding: ActivityOnboardingBinding
+
+  override fun onCreate(savedInstanceState: Bundle?) {
+    super.onCreate(savedInstanceState)
+
+    binding = ActivityOnboardingBinding.inflate(layoutInflater)
+    setContentView(binding.root)
+    val pagerAdapter = OnboardingPagerAdapter()
+    binding.onboardingPager.adapter = pagerAdapter
+    binding.onboardingPager.registerOnPageChangeCallback(object :
+        ViewPager2.OnPageChangeCallback() {
+      override fun onPageSelected(position: Int) {
+        binding.onboardingNextButton.text = if (position == NUM_PAGES - 1) {
+          getString(R.string.onboarding_done)
+        } else {
+          getString(R.string.onboarding_next)
+        }
+        binding.onboardingBackButton.visibility = if (position == 0) {
+          View.GONE
+        } else {
+          View.VISIBLE
+        }
+        super.onPageSelected(position)
+      }
+    })
+    binding.onboardingNextButton.setOnClickListener {
+      if (binding.onboardingPager.currentItem == NUM_PAGES - 1) {
+        MainActivity.startActivity(this)
+      } else {
+        binding.onboardingPager.currentItem = binding.onboardingPager.currentItem + 1
+      }
+    }
+    binding.onboardingBackButton.setOnClickListener {
+      binding.onboardingPager.currentItem = binding.onboardingPager.currentItem - 1
+    }
+
+    val sharedPref = PreferenceManager.getDefaultSharedPreferences(this)
+    with(sharedPref.edit()) {
+      putBoolean("onboarding", false)
+      apply()
+    }
+  }
+
+  override fun onBackPressed() {
+    if (binding.onboardingPager.currentItem == 0) {
+      // If the user is currently looking at the first step, allow the system to handle the
+      // Back button. This calls finish() on this activity and pops the back stack.
+      super.onBackPressed()
+    } else {
+      // Otherwise, select the previous step.
+      binding.onboardingPager.currentItem = binding.onboardingPager.currentItem - 1
+    }
+  }
+
+  private inner class OnboardingPagerAdapter :
+      RecyclerView.Adapter<OnboardingPagerAdapter.ViewHolder>() {
+
+    override fun getItemCount(): Int = NUM_PAGES
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+      return ViewHolder(
+          ItemOnboardingBinding.inflate(LayoutInflater.from(parent.context), parent, false))
+    }
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+      holder.bind(pages[position])
+    }
+
+    inner class ViewHolder(val itemBinding: ItemOnboardingBinding) :
+        RecyclerView.ViewHolder(itemBinding.root) {
+
+      fun bind(data: Pair<Int, Int>) {
+        itemBinding.onboardingImage.setImageResource(data.first)
+        itemBinding.onbardingText.setText(data.second)
+      }
+    }
+  }
+
+  companion object {
+    private const val NUM_PAGES = 4
```

Comment:
Any particular reason this can't/shouldn't be derived from the count of the `pages` list?